Std/Tactic/Case.lean

```diff
@@ -0,0 +1,148 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean.Elab.Tactic
+
+/-!
+# Extensions to the `case` tactic
+
+Adds a variant of `case` that looks for a goal with a particular type, rather than a goal
+with a particular tag.
+For consistency with `case`, it takes a tag as well, but the tag can be a hole `_`.
+-/
+
+namespace Std.Tactic
+open Lean Meta Elab Tactic
+
+/-- Clause for a `case ... : ...` tactic. -/
+syntax casePattArg := Parser.Tactic.caseArg (" : " term)?
+
+/-- The body of a `case ... : ...` tactic. -/
+syntax casePattBody := " => " (hole <|> syntheticHole <|> tacticSeq)
+
+/--
+* `case _ : t => tac` finds the first goal that unifies with `t` and then solves it
+  using `tac` or else fails. Like `show`, it changes the type of the goal to `t`.
+  The `_` can optionally be a case tag, in which case it only looks at goals
+  whose tag would be considered by `case` (goals with an exact tag match,
+  followed by goals with the tag as a suffix, followed by goals with the tag as a prefix).
+
+* `case _ n₁ ... nₘ : t => tac` additionally names the `m` most recent hypotheses with
+  inaccessible names to the given names. The names are renamed before matching against `t`.
+  The `_` can optionally be a case tag.
+
+* `case _ : t := e` is short for `case _ : t => exact e`.
+
+* `case _ : t₁ | _ : t₂ | ... => tac`
+  is equivalent to `(case _ : t₁ => tac); (case _ : t₂ => tac); ...`
+  but with all matching done on the original list of goals, in case pattern elaboration creates
+  new goals.
+  Each goal is consumed as they are matched, so patterns may repeat.
+
+* `case _ : t` will make the matched goal be the first goal.
+  `case _ : t₁ | _ : t₂ | ...` makes the matched goals be the first goals in the given order.
+
+* `case _ : t := _` and `case _ : t := ?m` are the same as `case _ : t` but in the `?m` case the
+  goal tag is changed to `m`.
+  In particular, the goal becomes metavariable `?m`.
+-/
+-- Low priority so that type-free `case` doesn't conflict with core `case`,
+-- though it should be a drop-in replacement.
+syntax (name := casePatt) (priority := low)
+  "case " sepBy1(casePattArg, " | ") (casePattBody)? : tactic
```

Comment:
```suggestion
  "case " sepBy1(casePattArg, " | ") (casePattTac <|> casePattExpr)? : tactic
```
where `casePattTac := " => " (hole <|> syntheticHole <|> tacticSeq)` and `casePattExpr := " := " colGt t:term`. I think this will behave better wrt the backtracking behavior of the parser compared to having two tactics with a long common prefix.